documentation/src/main/asciidoc/user_guide/chapter-63-Streams.adoc

```diff
@@ -0,0 +1,416 @@
+a sequential stream is forced locally.  The latter could be enhanced at some point to allow this.
+
+
+===== Remote requests
+When there are multiple nodes it may be desirable to control whether the remote requests are all processed
+at the same time concurrently or one at a time.  By default all terminal operations except the iterator
+perform concurrent requests.  The iterator method to reduce overall memory pressure on the local node
+only performs one at a time and actually performs slightly better.
+
+If a user wishes to change this default however they can do so by invoking the
+link:https://docs.jboss.org/infinispan/{infinispanversion}/apidocs/org/infinispan/CacheStream.html#sequentialDistribution--[sequentialDistribution]
+or link:https://docs.jboss.org/infinispan/{infinispanversion}/apidocs/org/infinispan/CacheStream.html#parallelDistribution--[parallelDistribution]
+methods on the `CacheStream`.   Note that currently intermediate operations return a `Stream` instance
+so you must make sure to invoke these methods before calling another intermediate operation.
+
+
+==== Task timeout
+
+It is possible to set a timeout value for the operation requests. This timeout is used only for remote requests timing out and
+it is on a per request basis. The former means the local execution will not timeout and the latter means if you have a failover
+scenario as described above the subsequent requests each have a new timeout.  If no timeout is specified, it uses the
+replication timeout as a default timeout. You can set the timeout in your task by doing the following: 
+
+[source,java]
+----
+CacheStream<Object, String> stream = cache.entrySet().stream();
+stream.timeout(1, TimeUnit.MINUTES);
+----
+
+For more information about this, please check the java doc in
+link:https://docs.jboss.org/infinispan/{infinispanversion}/apidocs/org/infinispan/CacheStream.html#timeout-long-java.util.concurrent.TimeUnit-[timeout]
+javadoc.
+
+==== Injection
+
+The link:https://docs.oracle.com/javase/8/docs/api/java/util/stream/Stream.html[Stream]
+has a terminal operation called
+link:https://docs.oracle.com/javase/8/docs/api/java/util/stream/Stream.html#forEach-java.util.function.Consumer-[forEach]
+which allows for running some sort of side effect operation on the data.  In this case it may be desirable to get a reference to
+the `Cache` that is backing this Stream.  In this case if your `Consumer` implements the 
+link:http://docs.jboss.org/infinispan/{infinispanversion}/apidocs/org/infinispan/stream/CacheAware.html[CacheAware]
+interface the `injectCache` method be invoked before the accept method from the `Consumer` interface.
+
+There are plans to possibly adding a new forEach that instead takes a
+link:https://docs.oracle.com/javase/8/docs/api/java/util/function/BiConsumer.html[BiConsumer]
+which would allow for much cleaner API usage since a lambda could be used.
```

Comment:
I wouldn't really mention this, but if you want to keep this paragraph, you should also link to an issue in JIRA so users can comment/vote for it.